rfc/rfc-0005-support-composing-supply-chains.md

```diff
@@ -1,3 +1,239 @@
+- The Snippet could explicitly define which component could emit values.
+
+_Could a Snippet emit multiple values?_ It should not. If a snippet emitted values
+from two ImageTemplates, for example, there would be no way for the calling
+SupplyChain to differentiate these two values (as the SupplyChain treats the Snippet
+as just another template).
+
+_Should snippets be typed?_ This would result in having a SupplyChainSourceSnippet,
+SupplyChainImageSnippet, SupplyChainConfigSnippet, etc. This multiplication of
+resources adds cognitive and code overhead in exchange for dubious value.
+
+### Example
+```yaml
+apiVersion: kontinue.io/v1alpha1
+kind: ClusterSupplyChainSnippet
+metadata:
+  name: responsible-ops-middle
+spec:
+  resources:
+    - name: built-image-provider
+      templateRef:
+        kind: ClusterImageTemplate
+        name: kpack-battery
+      sources:
+        - component: source-provider
+          name: solo-source-provider
+    - name: opinion-service-workload-template-provider
+      templateRef:
+        kind: ClusterConfigTemplate
+        name: opinion-service-battery
+      images:
+        - component: built-image-provider
+          name: solo-image-provider
+  exportedComponent:
+    name: opinion-service-workload-template-provider
+
+---
+
+apiVersion: kontinue.io/v1alpha1
+kind: ClusterSupplyChain
+metadata:
+  name: responsible-ops-composed
+spec:
+  selector:
+    integration-test: "my-workload"
+
+  resources:
+    - name: source-provider
+      templateRef:
+        kind: ClusterSourceTemplate
+        name: git-repository-battery
+
+    - name: composed-supply-chain-snippet
+      templateRef:
+        kind: ClusterSupplyChainSnippet
+        name: responsible-ops-middle
+      sources:
+        - component: source-provider
+          name: solo-source-provider
+
+    - name: cluster-sink
+      templateRef:
+        kind: ClusterTemplate
+        name: cluster-sink-battery
+      config:
+        - component: composed-supply-chain-snippet
+          name: singular-workload-template-provider
+
+```
+
+## Use Case
+
+RFC 9 introduced the ability for a resource in a supply chain to swap 1-1 between
+different templates, based on fields present on the workload. There is still
+need to allow a supply chain to swap 1-N templates, that is to allow an optional
+path through a supply chain to be of arbitrary length, unbound by the length of
+other paths. Snippets can allow this. A resource could choose between a regular
+template and a snippet. That snippet can be of arbitrary length.
+
+### Example
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSupplyChain
+metadata:
+  name: supply-chain
+spec:
+  selector:
+    matchLabels:
+      workload-type: web
+
+  resources:
+    - name: provide-image
+      templateRef:
+        kind: ClusterImageTemplate
+        options:
+          - name: building-image
+            selector:
+              matchFields:
+                or:
```

Comment:
I've mentioned that I expected boolean operators to become necessary at some point for our matchers. This is the first concrete use case I've had with them.

Reply:
the typically behavior for selectors in k8s is that each facet is ANDed together. If you need OR semantics, define multiple selectors. That won't work if you need nested boolean logic, but at the point the grammar is probably too complex anyway.

Reply:
Boolean operators are elidable by the fact that options are already "orable". Yes there's duplication, but its minimal and matches the intent of "when this, do this":

```yaml
 - name: provide-image
      templateRef:
        kind: ClusterImageTemplate
        options:
          - name: building-image
            selector:
              matchFields:
                - key: "spec.source.url"
                  operation: exists
          - name: building-image
            selector:
              matchFields:
                - key: "spec.source.image"
                  operation: exists
          - name: image-provider
            selector:
              matchFields:
                { key: "spec.image", operation: exists }
```

Reply:
Fair points! @scothis I'm curious if the example Rash gives here is what you had in mind.

Reply:
@waciumawanjohi that can certainly work, but my suggestion was more like

```yaml
 - name: provide-image
      templateRef:
        kind: ClusterImageTemplate
        options:
          - name: building-image
            selectors:
            - matchFields:
                - { key: "spec.source.url", operation: exists }
            - matchFields:
                - { key: "spec.source.image", operation: exists }
          - name: image-provider
            selectors:
            - matchFields:
                - { key: "spec.image", operation: exists }
```